### PR TITLE
feat(rnd,infra): Pull out websockets away from server api

### PR DESCRIPTION
**Pull request by aarushik93**
### Background
This is the base for splitting out into microservices. First is to purely keep as much of the same logic and code but pull out websocket server and rest server so they can be deployed seperately. 
As part of this also replaced the python queue with Redis. 

Eventually more of the in memory queue implementations can use redis. 
### Changes 🏗️
Added a Redis queue
Added a docker compose to run entire backend together
Added Dockerfile for websockets (Dockerfile.ws)
Added helm charts for websockets
### PR Quality Scorecard ✨

- [x] Have you used the PR description template? &ensp; `+2 pts`
- [ ] Is your pull request atomic, focusing on a single change? &ensp; `+5 pts`
- [ ] Have you linked the GitHub issue(s) that this PR addresses? &ensp; `+5 pts`
- [ ] Have you documented your changes clearly and comprehensively? &ensp; `+5 pts`
- [ ] Have you changed or added a feature? &ensp; `-4 pts`
  - [ ] Have you added/updated corresponding documentation? &ensp; `+4 pts`
  - [ ] Have you added/updated corresponding integration tests? &ensp; `+5 pts`
- [ ] Have you changed the behavior of AutoGPT? &ensp; `-5 pts`
  - [ ] Have you also run `agbenchmark` to verify that these changes do not regress performance? &ensp; `+10 pts`

### CODE REVIEW

File: rnd/infra/terraform/environments/dev.tfvars
Comment (aarushik93) [inline rnd/infra/terraform/environments/dev.tfvars]: we're not using the auth ip for now since we're using free tier of supabase. so repurposing it.

File: rnd/docker-compose.yml
Comment (aarushik93) [inline rnd/docker-compose.yml:60]: not real password

File: rnd/autogpt_builder/src/lib/autogpt-server-api/client.ts
Comment (majdyz) [inline rnd/autogpt_builder/src/lib/autogpt-server-api/client.ts]: Will the fallback even be possible if we host it as a different service? should we set the default value to a different port or route? by default

File: rnd/autogpt_server/autogpt_server/data/queue.py
Comment (majdyz) [inline rnd/autogpt_server/autogpt_server/data/queue.py:39]: Can we log a ticket or add a todo on ExecutionQueue on execution.py to use this ?

File: rnd/autogpt_server/autogpt_server/util/service.py
Comment (majdyz) [inline rnd/autogpt_server/autogpt_server/util/service.py]: Can we make this an option similar to `use_db` ? because I don't think all the service will use it.

File: rnd/autogpt_builder/src/lib/autogpt-server-api/client.ts
Comment (majdyz) [reply rnd/autogpt_builder/src/lib/autogpt-server-api/client.ts]: *8001

File: rnd/autogpt_server/pyproject.toml
Comment (majdyz) [inline rnd/autogpt_server/pyproject.toml:67]: I think we still need to have a script to run all of them in a single command for development purposes. Can we still make the ws_app.py an AppService ? and the app run both of them in a different port?

File: rnd/infra/helm/redis-values.dev.yaml
Comment (majdyz) [inline rnd/infra/helm/redis-values.dev.yaml:7]: Curious, why do we need persistence?

File: rnd/autogpt_server/pyproject.toml
Comment (aarushik93) [reply rnd/autogpt_server/pyproject.toml:67]: The plan was to go with docker compose and/or devcontainers for local development

File: rnd/autogpt_builder/src/lib/autogpt-server-api/client.ts
Comment (aarushik93) [reply rnd/autogpt_builder/src/lib/autogpt-server-api/client.ts]: > Will the fallback even be possible if we host it as a different service? should we set the default value to a different port or route? by default


what do you mean?

Comment (aarushik93) [reply rnd/autogpt_builder/src/lib/autogpt-server-api/client.ts]: updated to the websocket port

File: rnd/infra/helm/redis-values.dev.yaml
Comment (aarushik93) [reply rnd/infra/helm/redis-values.dev.yaml:7]: the k8s scheduler is free to move pods around on nodes or create on different nodes, this allows for moving the pods around while maintaining data

File: rnd/autogpt_server/Dockerfile.ws
Comment (Pwuts) [inline rnd/autogpt_server/Dockerfile.ws]: To reduce rebuild time and improve DX, could you reorder this so it doesn't have to re-install all dependencies when changing something in `rnd/autogpt_server`?

File: rnd/autogpt_server/Dockerfile.ws
Comment (Pwuts) [inline rnd/autogpt_server/Dockerfile.ws:1]: Since it's the same app with a different entrypoint (at least right now), it could be worth merging the Dockerfiles and defining multiple targets stacked on the same base image. Reduces duplication and makes maintenance easier.

File: rnd/autogpt_server/autogpt_server/server/ws_api.py
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/server/ws_api.py]: Copy-pasted from `.rest_api` -> better to move into a shared `.utils`

File: rnd/autogpt_server/autogpt_server/server/ws_api.py
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/server/ws_api.py:63]: What are these codes? Is there an overview of all of them and their meaning?

File: rnd/autogpt_server/autogpt_server/data/queue.py
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/data/queue.py]: This shouldn't be here

File: rnd/autogpt_server/autogpt_server/server/rest_api.py
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/server/rest_api.py]: Do we need to use `self.run_and_wait(..)`? The rest of `lifespan()` uses `await`.

File: rnd/autogpt_server/autogpt_server/server/ws_api.py
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/server/ws_api.py]: IIRC allowing CORS from any origin is a potential security risk (cc @ntindle)

File: rnd/autogpt_server/autogpt_server/server/ws_api.py
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/server/ws_api.py]: lint

File: rnd/autogpt_server/pyproject.toml
Comment (majdyz) [reply rnd/autogpt_server/pyproject.toml:67]: Shouldn't we have that first ready to prevent the awkward state of docker compose is not in the ready state for the devloop while the normal route becomes more complicated ?
Will docker-compose run the dev version of the app or the build version, will it auto-mount the file changes and does the hot-reloading etc.

**Review by majdyz — DISMISSED** [review]

File: rnd/autogpt_server/autogpt_server/server/rest_api.py
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/server/rest_api.py]: This is also called [here](https://github.com/Significant-Gravitas/AutoGPT/pull/7899/files#diff-0688cfa91a411a53076603b85abc3541971cd2de1d2812b4f467ddfed1eea443R75) on the `AppService`. Are they both needed?

File: rnd/autogpt_server/autogpt_server/util/service.py
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/util/service.py:49]: Just a note that this is generally a bad pattern as it will cause the `event_queue` instance to be shared across all `AppService` instances. In this specific case it probably won't cause any problems because every `AppService` runs in a separate process.

File: rnd/autogpt_server/pyproject.toml
Comment (Pwuts) [inline rnd/autogpt_server/pyproject.toml]: I don't understand why this is marked as diff because these lines are already in `master` 🤔

File: rnd/autogpt_server/autogpt_server/util/service.py
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/util/service.py]: Nit: It seems like most services don't need Redis, so maybe make this default `False` and enable it on the services that do need it.

File: rnd/autogpt_server/autogpt_server/data/queue.py
Comment (Pwuts) [inline rnd/autogpt_server/autogpt_server/data/queue.py]:
```suggestion
            logger.info(f"Connected to Redis on {self.host}:{self.port}")
```
same for the other occurrences in this file

File: rnd/docker-compose.yml
Comment (Pwuts) [inline rnd/docker-compose.yml]: Missing newline.

Tip: [NewLine VSCode extension](https://marketplace.cursorapi.com/items?itemName=chang196700.newline)
> Whenever you save a file, this extension will automatically add a blank line at its end, if there isn't one already. If you have multiple blank lines, only one will be kept.

File: rnd/infra/helm/autogpt-server/values.dev.yaml
Comment (Pwuts) [inline rnd/infra/helm/autogpt-server/values.dev.yaml:85]: Does this mean 100 node workers per graph worker? That seems excessive. If that's not what it means, I didn't say anything. :p

**Review by ntindle — DISMISSED** [review]

File: rnd/docker-compose.yml
Comment (Pwuts) [inline rnd/docker-compose.yml:1]: This will outdate the READMEs. Could you update those with the new (local dev) setup?

**Review by Pwuts — COMMENTED** [review]
Yee haw! Nice work. I have a few comments but nothing major, considering the size of this PR. Most important is to update the READMEs for the new `docker compose` setup. A heads-up about this in [#dev](https://discord.com/channels/1092243196446249134/1095817829405704305) once you merge would also be hugely appreciated. :)

File: rnd/docker-compose.yml
Comment (ntindle) [reply rnd/docker-compose.yml:1]: And if possible work with Toran on new videos

File: rnd/docker-compose.yml
Comment (ntindle) [reply rnd/docker-compose.yml]: Add this to the suggested extensions @Pwuts

File: rnd/autogpt_server/autogpt_server/server/ws_api.py
Comment (ntindle) [reply rnd/autogpt_server/autogpt_server/server/ws_api.py:63]: I’m pretty sure they’re similar to http but with an extra 0

File: rnd/autogpt_server/autogpt_server/server/ws_api.py
Comment (ntindle) [reply rnd/autogpt_server/autogpt_server/server/ws_api.py]: Yes, we will get roasted again for it I bet 😞

Comment (aarushik93) [reply rnd/autogpt_server/autogpt_server/server/ws_api.py]: I just copied over what was in the app. 
re security risk, it's already blocked on the ingress level we only allow specific domains so its not _actually_ a security risk
but we can also change it here too

File: rnd/autogpt_server/pyproject.toml
Comment (aarushik93) [reply rnd/autogpt_server/pyproject.toml]: i think something went wrong resolving conflicts. let me see

File: rnd/autogpt_server/autogpt_server/server/ws_api.py
Comment (ntindle) [reply rnd/autogpt_server/autogpt_server/server/ws_api.py]: For context on my message: We keep getting “roasted” by security researchers that don’t understand you can do it at different levels. 

I don’t actually care if you fix here because ya know, defense in depth should cover it by this point but if you don’t mind fixing it, I’d love to not have to triage a security vuln

File: rnd/autogpt_server/Dockerfile.ws
Comment (aarushik93) [reply rnd/autogpt_server/Dockerfile.ws:1]: I think the seperation makes it really clear, what's what especially when referencing in helm charts - ill log this as a ticket and if it becomes overhead we can change it

File: rnd/autogpt_server/autogpt_server/server/ws_api.py
Comment (aarushik93) [reply rnd/autogpt_server/autogpt_server/server/ws_api.py]: yep fixing it in this PR. my message was to also give context/knowledge share /peace of mind re security 
on this one

File: rnd/autogpt_server/autogpt_server/server/ws_api.py
Comment (aarushik93) [reply rnd/autogpt_server/autogpt_server/server/ws_api.py]: this isn't used here actually, removed it and put it in utils regardless

File: rnd/infra/helm/autogpt-server/values.dev.yaml
Comment (aarushik93) [reply rnd/infra/helm/autogpt-server/values.dev.yaml:85]: we're just seeing how far we can stretch the system ha

File: rnd/autogpt_server/autogpt_server/util/service.py
Comment (aarushik93) [reply rnd/autogpt_server/autogpt_server/util/service.py:49]: yeah i think its okay to leave for now though - they're all getting pulled out one by one

File: rnd/autogpt_server/autogpt_server/server/ws_api.py
Comment (Pwuts) [reply rnd/autogpt_server/autogpt_server/server/ws_api.py:63]: Then either the code or message on these is incorrect:
```py
await websocket.close(code=4002, reason="Invalid token")
```
and
```py
await websocket.close(code=4003, reason="Invalid token")
```

Comment (aarushik93) [reply rnd/autogpt_server/autogpt_server/server/ws_api.py:63]: the 400x are up to use to decide on how we want to use them 

we can eitther update the reasoning to make it more clear if we want. I've always just thought its better security to not give out exactly the reason ie invalid user_id

File: rnd/autogpt_server/Dockerfile.ws
Comment (Pwuts) [reply rnd/autogpt_server/Dockerfile.ws]: 1. Copy `pyproject.toml` and `poetry.lock`, and any path dependencies (so `forge` and `autogpt_libs`)
2. Install dependencies
3. Copy Prisma schema
4. `prisma generate`
5. Copy `rnd/autogpt_server/autogpt_server` source

File: rnd/autogpt_server/pyproject.toml
Comment (aarushik93) [reply rnd/autogpt_server/pyproject.toml:67]: im not sure what you mean by not in the ready state, I've added a readme now

but basically it stays mostly the same you just have to run the db and services via docker compose

and no, theres no hot-reloading

Comment (majdyz) [reply rnd/autogpt_server/pyproject.toml:67]: with no hot reloading, how do I add some changes, then try to test the change?
Do I need to do bash exec into the docker find the right service and restart it? Or do I need to `docker compose down` then `docker compose up` on each change ?

Comment (aarushik93) [reply rnd/autogpt_server/pyproject.toml:67]: To clarify then do you mean not in the ready state because of no hot reloading?

Comment (aarushik93) [reply rnd/autogpt_server/pyproject.toml:67]: As discussed on discord: 

Added the watch step, Nick is adding VScode config

**Comment by ntindle** [comment]
Got these during standup
```
rnd-ws_server-1    | Skipping virtualenv creation, as specified in config file.
rnd-ws_server-1    | Warning: 'ws' is an entry point defined in pyproject.toml, but it's not installed as a script. You may get improper `sys.argv[0]`.
rnd-ws_server-1    | 
rnd-ws_server-1    | The support to run uninstalled scripts will be removed in a future release.
rnd-ws_server-1    | 
rnd-ws_server-1    | Run `poetry install` to resolve and get rid of this message.
rnd-ws_server-1    | 
rnd-rest_server-1  | Warning: 'app' is an entry point defined in pyproject.toml, but it's not installed as a script. You may get improper `sys.argv[0]`.
rnd-rest_server-1  | 
rnd-rest_server-1  | The support to run uninstalled scripts will be removed in a future release.
rnd-rest_server-1  | 
rnd-rest_server-1  | Run `poetry install` to resolve and get rid of this message.
rnd-rest_server-1  | 
```

Also hit this when trying to create a graph from a template
```
rnd-rest_server-1  | INFO:     192.168.65.1:46167 - "POST /api/graphs HTTP/1.1" 500 Internal Server Error
rnd-rest_server-1  | ERROR:    Exception in ASGI application
rnd-rest_server-1  | Traceback (most recent call last):
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/uvicorn/protocols/http/httptools_impl.py", line 399, in run_asgi
rnd-rest_server-1  |     result = await app(  # type: ignore[func-returns-value]
rnd-rest_server-1  |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/uvicorn/middleware/proxy_headers.py", line 70, in __call__
rnd-rest_server-1  |     return await self.app(scope, receive, send)
rnd-rest_server-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/fastapi/applications.py", line 1054, in __call__
rnd-rest_server-1  |     await super().__call__(scope, receive, send)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/applications.py", line 123, in __call__
rnd-rest_server-1  |     await self.middleware_stack(scope, receive, send)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 186, in __call__
rnd-rest_server-1  |     raise exc
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/errors.py", line 164, in __call__
rnd-rest_server-1  |     await self.app(scope, receive, _send)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/cors.py", line 91, in __call__
rnd-rest_server-1  |     await self.simple_response(scope, receive, send, request_headers=headers)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/cors.py", line 146, in simple_response
rnd-rest_server-1  |     await self.app(scope, receive, send)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
rnd-rest_server-1  |     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
rnd-rest_server-1  |     raise exc
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
rnd-rest_server-1  |     await app(scope, receive, sender)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 758, in __call__
rnd-rest_server-1  |     await self.middleware_stack(scope, receive, send)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 778, in app
rnd-rest_server-1  |     await route.handle(scope, receive, send)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 299, in handle
rnd-rest_server-1  |     await self.app(scope, receive, send)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 79, in app
rnd-rest_server-1  |     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 64, in wrapped_app
rnd-rest_server-1  |     raise exc
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
rnd-rest_server-1  |     await app(scope, receive, sender)
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/starlette/routing.py", line 74, in app
rnd-rest_server-1  |     response = await func(request)
rnd-rest_server-1  |                ^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/fastapi/routing.py", line 299, in app
rnd-rest_server-1  |     raise e
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/fastapi/routing.py", line 294, in app
rnd-rest_server-1  |     raw_response = await run_endpoint_function(
rnd-rest_server-1  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/fastapi/routing.py", line 191, in run_endpoint_function
rnd-rest_server-1  |     return await dependant.call(**values)
rnd-rest_server-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/app/rnd/autogpt_server/autogpt_server/server/rest_api.py", line 314, in create_new_graph
rnd-rest_server-1  |     return await cls.create_graph(create_graph, is_template=False, user_id=user_id)
rnd-rest_server-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/app/rnd/autogpt_server/autogpt_server/server/rest_api.py", line 355, in create_graph
rnd-rest_server-1  |     return await graph_db.create_graph(graph, user_id=user_id)
rnd-rest_server-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/app/rnd/autogpt_server/autogpt_server/data/graph.py", line 399, in create_graph
rnd-rest_server-1  |     await __create_graph(tx, graph, user_id)
rnd-rest_server-1  |   File "/app/rnd/autogpt_server/autogpt_server/data/graph.py", line 410, in __create_graph
rnd-rest_server-1  |     await AgentGraph.prisma(tx).create(
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/prisma/actions.py", line 1208, in create
rnd-rest_server-1  |     resp = await self._client._execute(
rnd-rest_server-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/prisma/_base_client.py", line 533, in _execute
rnd-rest_server-1  |     return await self._engine.query(builder.build(), tx_id=self._tx_id)
rnd-rest_server-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/prisma/engine/_query.py", line 404, in query
rnd-rest_server-1  |     return await self.request(
rnd-rest_server-1  |            ^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/prisma/engine/_http.py", line 233, in request
rnd-rest_server-1  |     return self._process_response_data(data=data, response=response)
rnd-rest_server-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/prisma/engine/_http.py", line 87, in _process_response_data
rnd-rest_server-1  |     return utils.handle_response_errors(response, errors_data)
rnd-rest_server-1  |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rnd-rest_server-1  |   File "/usr/local/lib/python3.11/site-packages/prisma/engine/utils.py", line 175, in handle_response_errors
rnd-rest_server-1  |     raise exc(error)
rnd-rest_server-1  | prisma.errors.ForeignKeyViolationError: Foreign key constraint failed on the field: `foreign key`
```

**Comment by aarushik93** [comment]
@ntindle the first one is warnings, existing before this change and should affect actually running

number 2, taking a look have you run the migrations for postgres?

**Review by ntindle — APPROVED** [review]

**Review by ntindle — APPROVED** [review]

**Review by Pwuts — APPROVED** [review]
let's gooooooo

**Pull request merged**
